### PR TITLE
Add new method for checking error in CUDA Kernel

### PR DESCRIPTION
**Pull request by hbwx24**
### PR types

New features
### PR changes

Others
### Describe

Add new method for checking error in CUDA Kernel

This PR  checks for errors in the cuda kernel through **PADDLE_ENFORCE_CUDA_KERNEL** ， and then  issues an error message through **PADDLE_ENFORCE_CHECK_CUDA_KERNEL**.The added code does not work in the case of -DWITH_LITE=ON or OpenBLAS.The modification of the code for checking errors only involves **nll_loss_op**, **bernoulli_op** and **lookup_table_v2_op**, and the rest of the operators will be modified later.

#### Example

Original error:
```
Error: /root/codes/Paddle/paddle/fluid/operators/bernoulli_op.cu:36 Assertion `p >= 0.0 && p <= 1.0` failed. The probability should be >=0 and <= 1, but got 1.142277
Error: /root/codes/Paddle/paddle/fluid/operators/bernoulli_op.cu:36 Assertion `p >= 0.0 && p <= 1.0` failed. The probability should be >=0 and <= 1, but got 1.858263
Error: /root/codes/Paddle/paddle/fluid/operators/bernoulli_op.cu:36 Assertion `p >= 0.0 && p <= 1.0` failed. The probability should be >=0 and <= 1, but got 1.208680
Traceback (most recent call last):
  File "paddle-bernoll.py", line 14, in <module>
    print(out)
  File "/usr/local/python3.5.1/lib/python3.5/site-packages/paddle/fluid/dygraph/varbase_patch_methods.py", line 252, in __str__
    return 'Tensor: %s\n%s' % (self.name, str(tensor))
paddle.fluid.core_avx.EnforceNotMet: 

--------------------------------------
C++ Traceback (most recent call last):
--------------------------------------
0   paddle::framework::operator<<(std::ostream&, paddle::framework::LoDTensor const&)
1   paddle::framework::operator<<(std::ostream&, paddle::framework::Tensor const&)
2   paddle::framework::TensorCopy(paddle::framework::Tensor const&, paddle::platform::Place const&, paddle::platform::DeviceContext const&, paddle::framework::Tensor*)
3   void paddle::memory::Copy<paddle::platform::CPUPlace, paddle::platform::CUDAPlace>(paddle::platform::CPUPlace, void*, paddle::platform::CUDAPlace, void const*, unsigned long, CUstream_st*)
4   paddle::platform::GpuMemcpyAsync(void*, void const*, unsigned long, cudaMemcpyKind, CUstream_st*)
5   paddle::platform::EnforceNotMet::EnforceNotMet(std::__exception_ptr::exception_ptr, char const*, int)
6   std::string paddle::platform::GetTraceBackString<char const*>(char const*&&, char const*, int)
7   paddle::platform::GetCurrentTraceBackString()

----------------------
Error Message Summary:
----------------------
ExternalError:  Cuda error(4), unspecified launch failure.
  [Advise: An exception occurred on the device while executing a kernel. Common causes include dereferencing an invalid device pointerand accessing out of bounds shared memory. The device cannot be used until cudaThreadExit() is called. All existing device memory allocations are invalid and must be reconstructed if the program is to continue usingCUDA.] (at /root/codes/Paddle/paddle/fluid/platform/gpu_info.cc:286)
```

New format:
```
Traceback (most recent call last):
  File "paddle-bernoll.py", line 13, in <module>
    out = paddle.bernoulli(x)
  File "/usr/local/python3.5.1/lib/python3.5/site-packages/paddle/tensor/random.py", line 77, in bernoulli
    return core.ops.bernoulli(x)
paddle.fluid.core_avx.EnforceNotMet: 

--------------------------------------
C++ Traceback (most recent call last):
--------------------------------------
0   paddle::imperative::Tracer::TraceOp(std::string const&, paddle::imperative::NameVarBaseMap const&, paddle::imperative::NameVarBaseMap const&, paddle::framework::AttributeMap)
1   paddle::imperative::Tracer::TraceOp(std::string const&, paddle::imperative::NameVarBaseMap const&, paddle::imperative::NameVarBaseMap const&, paddle::framework::AttributeMap, paddle::platform::Place const&, bool)
2   paddle::imperative::OpBase::Run(paddle::framework::OperatorBase const&, paddle::imperative::NameVarBaseMap const&, paddle::imperative::NameVarBaseMap const&, paddle::framework::AttributeMap const&, paddle::platform::Place const&)
3   paddle::imperative::PreparedOp::Run(paddle::imperative::NameVarBaseMap const&, paddle::imperative::NameVarBaseMap const&, paddle::framework::AttributeMap const&)
4   std::_Function_handler<void (paddle::framework::ExecutionContext const&), paddle::framework::OpKernelRegistrarFunctor<paddle::platform::CUDAPlace, false, 0ul, paddle::operators::BernoulliOpKernel<paddle::platform::CUDADeviceContext, float>, paddle::operators::BernoulliOpKernel<paddle::platform::CUDADeviceContext, double> >::operator()(char const*, char const*, int) const::{lambda(paddle::framework::ExecutionContext const&)#1}>::_M_invoke(std::_Any_data const&, paddle::framework::ExecutionContext const&)
5   paddle::operators::BernoulliOpKernel<paddle::platform::CUDADeviceContext, float>::Compute(paddle::framework::ExecutionContext const&) const
6   paddle::platform::EnforceNotMet::EnforceNotMet(std::string const&, char const*, int)
7   paddle::platform::GetCurrentTraceBackString()

----------------------
Error Message Summary:
----------------------
OutOfRangeError:The probability should be >=0 and <= 1 (at /root/codes/Paddle/paddle/fluid/operators/bernoulli_op.cu:37)
  [operator < bernoulli > error]
```
- The original code sends a lot of repeated assertion information. 
- Tt is difficult for users to find the cause of the error based on the Error Message Summary.
#### TODO

- Apply to all operators
- Reduce the impact of the added code on other parts
- Support WITH_LITE=ON and OpenBLAS

### CODE REVIEW

File: paddle/fluid/operators/bernoulli_op.cu [hunk 1]
```diff
@@ -32,8 +33,10 @@ struct BernoulliCudaFunctor {
   __host__ __device__ T operator()(const unsigned int n, const T p) const {
     // NOTE(zhiqiu): currently, PADDLE_ENFORCE in cuda kernel may print several
     // lines of error messages if, and it should be refined.
-    PADDLE_ENFORCE(p >= 0.0 && p <= 1.0,
-                   "The probability should be >=0 and <= 1, but got %f", p);
+    PADDLE_ENFORCE_CUDA_KERNEL(p >= 0.0 && p <= 1.0,
+                               "The probability should be >=0 and <= 1");
+    // PADDLE_ENFORCE(p >= 0.0 && p <= 1.0,
```
Comment (chenwhql) [inline paddle/fluid/operators/bernoulli_op.cu]: remove useless comments

File: paddle/fluid/operators/bernoulli_op.cu [hunk 2]
```diff
@@ -32,8 +33,10 @@ struct BernoulliCudaFunctor {
   __host__ __device__ T operator()(const unsigned int n, const T p) const {
     // NOTE(zhiqiu): currently, PADDLE_ENFORCE in cuda kernel may print several
     // lines of error messages if, and it should be refined.
-    PADDLE_ENFORCE(p >= 0.0 && p <= 1.0,
-                   "The probability should be >=0 and <= 1, but got %f", p);
+    PADDLE_ENFORCE_CUDA_KERNEL(p >= 0.0 && p <= 1.0,
+                               "The probability should be >=0 and <= 1");
```
Comment (chenwhql) [inline paddle/fluid/operators/bernoulli_op.cu:38]: why remove `but got %f`, can't it be used here?

File: paddle/fluid/operators/lookup_table_v2_op.cu [hunk 3]
```diff
@@ -31,16 +32,20 @@ __global__ void LookupTableV2(T *output, const T *table, const int64_t *ids,
...
         "Variable value (input) of OP(fluid.layers.embedding) "
-        "expected >= 0 and < %ld, but got %ld. Please check input value.",
-        N, id);
-    PADDLE_ENFORCE(
-        id < N,
-        "Variable value (input) of OP(fluid.layers.embedding) "
-        "expected >= 0 and < %ld, but got %ld. Please check input value.",
-        N, id);
+        "expected >= 0 and < N.");
+    // PADDLE_ENFORCE(
```
Comment (chenwhql) [inline paddle/fluid/operators/lookup_table_v2_op.cu]: same above

File: paddle/fluid/operators/lookup_table_v2_op.cu [hunk 4]
```diff
@@ -31,16 +32,20 @@ __global__ void LookupTableV2(T *output, const T *table, const int64_t *ids,
...
+        id >= 0 && id < N,
         "Variable value (input) of OP(fluid.layers.embedding) "
-        "expected >= 0 and < %ld, but got %ld. Please check input value.",
-        N, id);
-    PADDLE_ENFORCE(
-        id < N,
-        "Variable value (input) of OP(fluid.layers.embedding) "
-        "expected >= 0 and < %ld, but got %ld. Please check input value.",
-        N, id);
+        "expected >= 0 and < N.");
```
Comment (chenwhql) [inline paddle/fluid/operators/lookup_table_v2_op.cu:38]: same above,

File: paddle/fluid/platform/cuda_enforce.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+#include "paddle/fluid/platform/enforce.h"
+#ifdef PADDLE_WITH_CUDA
+#ifdef PADDLE_CUDA_KERNEL_CHECK
+
+#include <cuda.h>
+#include <stdio.h>
+
+#define __LEN_ERROR_MSG 128
+
+#define PADDLE_ENFORCE_CHECK_CUDA_KERNEL()                               \
```
Comment (chenwhql) [inline paddle/fluid/platform/cuda_enforce.h]: It is recommended to explain the usage and limit of the new macro, avoid misuse

File: paddle/fluid/platform/cuda_enforce.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      return;                                                \
+    }                                                        \
+  } while (0)
+
+__device__ __host__ void send_error_msg(char* msg, char* file, int line);
+
+char ifCudaFail();
+void get_msg_from_cuda(char* msg, char* file, int* line);
+
+#define PADDLE_ENFORCE_CUDA_KERNEL _PADDLE_ENFORCE_CUDA_KERNEL
```
Comment (chenwhql) [inline paddle/fluid/platform/cuda_enforce.h:88]: same above

File: paddle/fluid/platform/cuda_enforce.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+#define PADDLE_ENFORCE_CHECK_CUDA_KERNEL()                               \
+  do {                                                                   \
+    if (UNLIKELY(ifCudaFail())) {                                        \
+      char msg[__LEN_ERROR_MSG] = {0};                                   \
+      int line = 0;                                                      \
+      char file[__LEN_ERROR_MSG] = {0};                                  \
+      get_msg_from_cuda(msg, file, &line);                               \
+      char err_msg_temp[__LEN_ERROR_MSG + 32] = {};                      \
+      snprintf(err_msg_temp, __LEN_ERROR_MSG, "Error:%s", msg);          \
```
Comment (chenwhql) [inline paddle/fluid/platform/cuda_enforce.h]: insert blank between `Error` and `:`, and maybe we can use `InvalidArgumentError` directly

File: paddle/fluid/platform/cuda_enforce.h (hunk 5 above)
Comment (chenwhql) [reply paddle/fluid/platform/cuda_enforce.h]: ref like: https://github.com/PaddlePaddle/Paddle/blob/7cb4a8b8f21b3432ceeacdc23ddf8b6841cf0e50/paddle/fluid/platform/enforce.h#L436

File: paddle/fluid/operators/bernoulli_op.cu (hunk 1 above)
Comment (hbwx24) [reply paddle/fluid/operators/bernoulli_op.cu]: Done, thx.

File: paddle/fluid/operators/bernoulli_op.cu (hunk 2 above)
Comment (hbwx24) [reply paddle/fluid/operators/bernoulli_op.cu:38]: 在device函数中，可以使用printf，但是不能用snprintf

File: paddle/fluid/platform/cuda_enforce.h (hunk 5 above)
Comment (hbwx24) [reply paddle/fluid/platform/cuda_enforce.h]: done, thx.

File: paddle/fluid/platform/cuda_enforce.h (hunk 7 above)
Comment (hbwx24) [reply paddle/fluid/platform/cuda_enforce.h]: done, thx.

**Pull request closed without merging**